app/controllers/events_controller.rb

```diff
@@ -1,68 +1,87 @@
+  rescue ActiveRecord::RecordInvalid
+    render json: { error: 'neither event_id nor bridge_id were valid' }, status: 400 # Bad Request
+  rescue ActiveRecord::RecordNotFound
+    render json: { error: 'events matching that id were not found' }, status: 400
+  rescue ActiveRecord::NotNullViolation
+    render json: { error: 'neither event_id nor bridge_id were not submitted' }, status: 400 # Bad Request
+  end
+
+  def show
+    event = Event.find(event_params[:id])
+    render json: event, status: 200
+  rescue ActiveRecord::RecordNotFound
+    render json: { error: 'an event by that id was not found' }, status: 400
+  end
+
+  def destroy
+    event = Event.find(event_params[:id])
+    event.destroy
+  end
 
-  # receive bridge id + data
   def create
-    bridge = Bridge.find(params[:id])
-    data = { inbound: request.body, outbounds: [] }
-    event = Event.new(data: data, bridge_id: bridge.id)
-    if event.save
-      # EventWorker.perform(id of event just created)
-      status 201 # Created
-    else
-      status 400 # Bad Request
-    end
+    event = create_event_object(create_data_object, find_bridge)
+    event.save!
+    EventWorker.perform_async(event.id)
+    render json: {}, status: 202 # Accepted (asynchronous processing)
+  rescue ActiveRecord::RecordNotFound
+    render json: { error: 'a bridge by that id was not found' }, status: 400
+  rescue ActiveRecord::RecordInvalid
+    render json: { error: 'payload, bridge_id, or urls were invalid' }, status: 400 # Bad Request
+  rescue ActiveRecord::NotNullViolation
+    render json: { error: 'payload, bridge_id, or urls fields were not submitted' }, status: 400 # Bad Request
   end
 
   private
 
-  def date_format(_date)
-    year = time.split('-')[0]
-    month = time.split('-')[1]
-    day = time.split('-')[2]
-    "#{year}-#{month}-#{day}"
+  def event_params
+    params.permit(:id, :bridge_id, :event_id, :test)
   end
 
-  def set_user
-    # bridge_id or #event_id
-    # => User
+  def create_data_object
+    datetime = DateTime.now.utc.to_s.split(' ')
+    { 'inbound' => {
+      'payload' => JSON.parse(request.body.read),
+      'date' => datetime.first,
+      'time' => datetime[1],
+      'ip' => request.ip,
+      'content_length' => request.content_length
+    },
+      'outbound' => [] }
   end
-end
 
-# Step 1
-# Convert binary to jsonb
+  def create_event_object(data, bridge)
+    Event.new(
+      data: data.to_json,
+      bridge_id: bridge.id,
+      test: event_params[:test] || false
+    )
+  end
 
-# Step 2
-# payload:
-{
-  test: 'user entered string from editor',
-  production: 'user entered string from editor'
-}
+  def create_sidebar_events(events)
+    events.map(&:sidebar_format).sort_by { |event| event[:id] }.reverse
+  end
 
-# Step 3
-#
+  def retrieve_events
+    if event_params[:bridge_id]
+      Bridge.find(event_params[:bridge_id]).events
+    elsif event_params[:event_id]
+      Event.find(event_params[:event_id]).bridge.events
```

Comment:
These should be using the includes query method.

https://apidock.com/rails/ActiveRecord/QueryMethods/includes

Reply:
Ok, I'll work on a more efficient way of loading these.